rpc/swap.go

```diff
@@ -46,37 +46,68 @@ func NewSwapService(
 	}
 }
 
-// PastSwapInfo contains a swap ID and its start and end time.
-type PastSwapInfo struct {
-	ID        types.Hash `json:"id" validate:"required"`
-	StartTime time.Time  `json:"startTime" validate:"required"`
-	EndTime   *time.Time `json:"endTime"`
+// PastSwap represents a past swap returned by swap_getPast.
+type PastSwap struct {
+	ID             types.Hash          `json:"id" validate:"required"`
+	Provided       coins.ProvidesCoin  `json:"provided" validate:"required"`
+	ProvidedAmount *apd.Decimal        `json:"providedAmount" validate:"required"`
+	ExpectedAmount *apd.Decimal        `json:"expectedAmount" validate:"required"`
+	ExchangeRate   *coins.ExchangeRate `json:"exchangeRate" validate:"required"`
+	Status         types.Status        `json:"status" validate:"required"`
+	StartTime      time.Time           `json:"startTime" validate:"required"`
+	EndTime        *time.Time          `json:"endTime"`
+}
+
+// GetPastRequest ...
+type GetPastRequest struct {
+	OfferID types.Hash `json:"offerID"`
 }
 
-// GetPastIDsResponse ...
-type GetPastIDsResponse struct {
-	Swaps []*PastSwapInfo `json:"swaps" validate:"dive,required"`
+// GetPastResponse ...
+type GetPastResponse struct {
+	Swaps []*PastSwap `json:"swaps" validate:"required"`
```

Comment:
This validation is saying that the slice must be non-empty. Can it be empty? I suspect that you wanted `dive,required` to validate that each entry in the slice is non-nil. If you wanted to validate that both the slice is non-empty and that each pointer in the slice is non-nil, you would use `required,dive,required`.

Reply:
updated